### PR TITLE
[Arrows] MC-21844: SVC false-positive: short names & FQCN in DockBlock

### PR DESCRIPTION
**Pull request by zakdma**
### Bug - P1
* [MC-21844](https://jira.corp.magento.com/browse/MC-21844) SVC false-positive: short names & FQCN in DockBlock

### Related Pull Requests

Created by : @zakdma

### CODE REVIEW

**Comment by zakdma** [comment]
@roribio could you please take a look at our PR

File: src/Visitor/NameResolver.php [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+class NameResolver extends ParserNameResolver
+{
+    /**
+     * @inheritDoc
+     */
+    public function enterNode(Node $node)
+    {
+        $return = parent::enterNode($node);
+
+        if ($node instanceof ClassMethod) {
```
Comment (vkublytskyi) [inline src/Visitor/NameResolver.php:37]: `Property`, `ClassConst` also may have annotation that describes a type. For Magento this is a less important cases and may be addressed in separate PRs.

File: src/Visitor/NameResolver.php [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+     * @return void
+     */
+    private function resolveDocBlockParamTypes(ClassMethod $node)
+    {
+        /** @var PhpDocNode $docNode */
+        $docNode = $this->getParsedDocNode($node);
+        if ($docNode) {
+            $result = [];
+            /** @var PhpDocTagNode[] $paramTags */
+            $paramTags = $docNode->getTagsByName('@param');
```
Comment (vkublytskyi) [inline src/Visitor/NameResolver.php:55]: Note: `getParamTagValues` returns node with more suitable API

File: src/Visitor/NameResolver.php [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+            foreach ($paramTags as $paramTag) {
+                $paramNode = [
+                    'name' => $paramTag->value->parameterName ?? '',
+                    'type' => $this->parseType($paramTag->value->type),
+                ];
+                $result['params'][] = $paramNode;
+            }
+
+            /** @var PhpDocTagNode[] $returnTags */
+            $returnTags = $docNode->getTagsByName('@return');
```
Comment (vkublytskyi) [inline src/Visitor/NameResolver.php:66]: Note: `getReturnTagValues` returns node with more suitable API

File: src/Visitor/NameResolver.php [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+    private function parseType($type)
+    {
+        $result = [];
+        if ($type instanceof UnionTypeNode) {
+            foreach ($type->types as $typeNode) {
+                $normalizedType = BuilderHelpers::normalizeType((string)$typeNode);
+                $resolvedType = $this->resolveType($normalizedType);
+                $result[] = $resolvedType;
+            }
+        } else {
```
Comment (vkublytskyi) [inline src/Visitor/NameResolver.php:91]: Note: Please ensure that `PHPStan\PhpDocParser\Ast\Type\NullableTypeNode` should not be handled as special case (equals to union `null|type`)

**Review by vkublytskyi — APPROVED** [review]

**Review by vkublytskyi — CHANGES_REQUESTED** [review]
Please fix unit test

File: src/Visitor/NameResolver.php (hunk 2 above)
Comment (zakdma) [reply src/Visitor/NameResolver.php]: fixed

File: src/Visitor/NameResolver.php (hunk 3 above)
Comment (zakdma) [reply src/Visitor/NameResolver.php]: fixed

File: src/Visitor/NameResolver.php (hunk 4 above)
Comment (zakdma) [reply src/Visitor/NameResolver.php:92]: ?type is parsed as NullableType
type|null is parsed as UnionType
So I handle only UnionType case as list of FullyQualified types because there can be also annotation like that
Product|int|null

**Comment by zakdma** [comment]
@vkublytskyi all fixed except constants and variable types parsing
It can be done in separate fix for that cases

**Review by roribio — APPROVED** [review]
Approved 💯

**Review by vkublytskyi — APPROVED** [review]
